File: LeetCode/Study Plan/Practice Problems/Medium/Walls And Gates/Go/walls_and_gates.go
package main

import (
	"fmt"
)

type Pair struct{ x, y, dis int }

func walls_and_gate(rooms [][]int) [][]int {
	queue := []Pair{}
	n, m := len(rooms), len(rooms[0])
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			if rooms[i][j] == 0 {
				queue = append(queue, Pair{x: i, y: j, dis: 0})
			}
		}
	}
	visited := make([][]bool, n)
	for i := 0; i < n; i++ {
		visited[i] = make([]bool, m)
	}

	for len(queue) > 0 {
		i, j, dis := queue[0].x, queue[0].y, queue[0].dis
		visited[i][j] = true
		queue = queue[1:]
		if (i+1 < n && j < m) && rooms[i+1][j] != -1 {
			if rooms[i+1][j] == 0 && !visited[i+1][j] {
				queue = append(queue, Pair{i + 1, j, 0})
			} else {
				if rooms[i+1][j] > dis+1 {
					rooms[i+1][j] = dis + 1
					queue = append(queue, Pair{i + 1, j, dis + 1})
				}
			}
		}
		if (i-1 >= 0 && j < m) && rooms[i-1][j] != -1 {
			if rooms[i-1][j] == 0 && !visited[i-1][j] {
				queue = append(queue, Pair{i - 1, j, 0})
			} else {
				if rooms[i-1][j] > dis+1 {
					rooms[i-1][j] = dis + 1
					queue = append(queue, Pair{i - 1, j, dis + 1})
				}
			}
		}
		if (i < n && j+1 < m) && rooms[i][j+1] != -1 {
			if rooms[i][j+1] == 0 && !visited[i][j+1] {
				queue = append(queue, Pair{i, j + 1, 0})
			} else {
				if rooms[i][j+1] > dis+1 {
					rooms[i][j+1] = dis + 1
					queue = append(queue, Pair{i, j + 1, dis + 1})
				}
			}
		}
		if (i < n && j-1 >= 0) && rooms[i][j-1] != -1 {
			if rooms[i][j-1] == 0 && !visited[i][j-1] {
				queue = append(queue, Pair{i, j - 1, 0})
			} else {

				if rooms[i][j-1] > dis+1 {
					rooms[i][j-1] = dis + 1
					queue = append(queue, Pair{i, j - 1, dis + 1})
				}
			}
		}
	}

	return rooms
}

func main() {
	fmt.Println(walls_and_gate([][]int{
		{2147483647, -1, 0, 2147483647},
		{2147483647, 2147483647, 2147483647, -1},
		{2147483647, -1, 2147483647, -1},
		{0, -1, 2147483647, 2147483647},
	}))
}

File: LeetCode/Medium/Walls And Gates/c++/walls_and_gates.cc
class Solution {
public:
    void wallsAndGates(vector<vector<int>>& rooms) {
        int m = rooms.size();
        int n = rooms[0].size();
        vector<vector<bool>> visited(m, vector<bool>(n, false));
        queue<vector<int>> q;
        
        for (int r = 0; r < m; r++) {
            for (int c = 0; c < n; c++) {
                if (rooms[r][c] == 0) q.push({r, c, 0});
            }
        }
        vector<vector<int>> dir = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        while (!q.empty()) {
            int size = q.size();
            auto curr_ele = q.front();
            int row = curr_ele[0];
            int col = curr_ele[1];
            int dis = curr_ele[2];
            q.pop();
            visited[row][col] = true;
            for (int d = 0; d < 4; d++) {
                int new_row = row + dir[d][0];
                int new_col = col + dir[d][1];
                if (is_safe(new_row, new_col, m, n, visited, rooms)) {
                    rooms[new_row][new_col] = dis + 1;
                    visited[new_row][new_col] = true;
                    q.push({new_row, new_col, dis + 1});
                }
            }
        }
    }

    bool is_safe(int row, int col, int m, int n, vector<vector<bool>>& visited, vector<vector<int>>& rooms) {
        return !(row < 0 || col < 0 || row >= m || col >= n || visited[row][col] || rooms[row][col] == -1 || rooms[row][col] == 0 || rooms[row][col] != INT_MAX);
    }
};
